src/soc/src/serial.rs

```diff
@@ -0,0 +1,26 @@
+use core::fmt;
+
+use embedded_hal_nb::serial::{ErrorType, Write};
+use nb::block;
+
+mod error;
+
+pub use error::Error;
+
+/// Generic trait for access to SoC serial peripheral(s).
+pub trait Serial: ErrorType + Write {}
+
+impl fmt::Write for dyn Serial<Error = Error> {
+    fn write_str(&mut self, s: &str) -> Result<(), fmt::Error> {
+        for &byte in s.as_bytes() {
+            // Inject a carriage return before a newline
+            if byte == b'\n' {
+                block!(self.write(b'\r')).ok();
+            }
+            block!(self.write(byte)).ok();
+        }
+        #[cfg(not(feature = "jh71xx"))]
```

Comment:
Missed removing this when I moved the implementation. Will remove with the next set of edits.